### ISSUE TITLE
[bug #28108] bug with winavr20090313

### ISSUE THREAD

**Issue by avrs-admin** [avrdudes/avr-libc#401]
Fri 27 Nov 2009 07:40:01 PM CET

greetings,

I'm using winavr20090313 which has gcc 4.3.2 and i'm getting the bug that I reported at:
 [http://www.avrfreaks.net/index.php?name=PNphpBB2&file=viewtopic&t=86733](https://savannah.nongnu.orghttp://www.avrfreaks.net/index.php?name=PNphpBB2&file=viewtopic&t=86733)

In attachment there's a .rar that shows the error.

Give me a help on this one please.

Thanks ;)

[file #19140: can_bug.rar](https://savannah.nongnu.org/bugs/download.php?file_id=19140)
[file #19155: can_bug.zip](https://savannah.nongnu.org/bugs/download.php?file_id=19155)

This issue was migrated from https://savannah.nongnu.org/bugs/?28108

**Comment by avrs-admin** [comment]
Brn Gomes <unknwn>
Mon 30 Nov 2009 12:17:46 PM CET

If I want a CAN msg with ID=0xFDFFFFF0 on msg2

it sends

ID=0xFDFDFFF0....

#define TEST_CAN_MSG1_ID 0xFDFFFFE0L
#define TEST_CAN_MSG2_ID 0xFDFFFFD0L
#define TEST_CAN_MSG3_ID 0xFDFFFFC0L

#define TEST_CAN_TX_MSG0_ID TEST_CAN_MSG3_ID
#define TEST_CAN_TX_MSG1_ID TEST_CAN_MSG1_ID
#define TEST_CAN_TX_MSG2_ID TEST_CAN_MSG2_ID

typedef struct
{
uint32_t id;
uint32_t id_msk;
uint8_t *data_ptr;
uint8_t status;
uint8_t proc;
void *action_ptr;
uint8_t max_size;
uint8_t size;
uint8_t mob_num;
uint8_t wdlc;
uint8_t mob_assign;
} can_msg_t;

static volatile can_msg_t can_tx_msg0;
static volatile can_msg_t can_tx_msg1;
static volatile can_msg_t can_tx_msg2;

static volatile uint8_t can_buf_out_msg0[10];
static volatile uint8_t can_buf_out_msg1[10];
static volatile uint8_t can_buf_out_msg2[10];

funtion called in C code:

void InitCan(void){

/*

- Msg 0*/
can_tx_msg0.id = TEST_CAN_TX_MSG0_ID;
can_tx_msg0.data_ptr = (uint8_t *) can_buf_out_msg0;
can_tx_msg0.max_size = 8;
can_tx_msg0.size = 8;
can_tx_msg0.status = NOT_SENT;
can_tx_msg0.mob_num = 0;
can_tx_msg0.mob_assign = FIXED_MOB;

/* send dummy data... */
can_buf_out_msg0[0] = 0;
can_buf_out_msg0[1] = 0;
can_buf_out_msg0[2] = 0;
can_buf_out_msg0[3] = 0;
can_buf_out_msg0[4] = 0xDE;
can_buf_out_msg0[5] = 0xAD;
can_buf_out_msg0[6] = 0xFA;
can_buf_out_msg0[7] = 0xCE;

/*

- Msg 1*/
can_tx_msg1.id = TEST_CAN_TX_MSG1_ID;
can_tx_msg1.data_ptr = (uint8_t *) can_buf_out_msg1;
can_tx_msg1.max_size = 8;
can_tx_msg1.size = 8;
can_tx_msg1.status = NOT_SENT;
can_tx_msg1.mob_assign = FLOATING_MOB;

/* send dummy data... */
can_buf_out_msg1[0] = 0xDE;
can_buf_out_msg1[1] = 0xAD;
can_buf_out_msg1[2] = 0xFA;
can_buf_out_msg1[3] = 0xCE;
can_buf_out_msg1[4] = 0;
can_buf_out_msg1[5] = 0;
can_buf_out_msg1[6] = 0;
can_buf_out_msg1[7] = 0;

/*

- Msg 2*/
can_tx_msg2.id = TEST_CAN_TX_MSG2_ID;
can_tx_msg2.data_ptr = (uint8_t *) can_buf_out_msg2;
can_tx_msg2.max_size = 8;
can_tx_msg2.size = 8;
can_tx_msg2.status = NOT_SENT;
can_tx_msg2.mob_assign = FLOATING_MOB;

/* send dummy data... */
can_buf_out_msg2[0] = 0xDE;
can_buf_out_msg2[1] = 0xAD;
can_buf_out_msg2[2] = 0xFA;
can_buf_out_msg2[3] = 0xCE;
can_buf_out_msg2[4] = 0;
can_buf_out_msg2[5] = 0;
can_buf_out_msg2[6] = 0;
can_buf_out_msg2[7] = 0;

......-----------......

the .lss file generated with the error at line 724:

can_tx_msg2.id = TEST_CAN_TX_MSG2_ID;
71a:	00 ed ldi	r16, 0xD0	; 208
71c:	e0 2e mov	r14, r16
71e:	0f ef ldi	r16, 0xFF	; 255
720:	f0 2e mov	r15, r16
722:	0f ef ldi	r16, 0xFF	; 255
724:	00 2f mov	r16, r16
726:	0d ef ldi	r16, 0xFD	; 253
728:	10 2f mov	r17, r16
72a:	e0 92 0f 02 sts	0x020F, r14
72e:	f0 92 10 02 sts	0x0210, r15
732:	00 93 11 02 sts	0x0211, r16
736:	10 93 12 02 sts	0x0212, r17
can_tx_msg2.data_ptr = (uint8_t *) can_buf_out_msg2;
73a:	86 e3 ldi	r24, 0x36	; 54
73c:	92 e0 ldi	r25, 0x02	; 2
73e:	90 93 18 02 sts	0x0218, r25
742:	80 93 17 02 sts	0x0217, r24
can_tx_msg2.max_size = 8;
746:	60 93 1d 02 sts	0x021D, r22
can_tx_msg2.size = 8;
74a:	60 93 1e 02 sts	0x021E, r22
can_tx_msg2.status = NOT_SENT;
74e:	70 93 19 02 sts	0x0219, r23

**Comment by avrs-admin** [comment]
Brn Gomes <unknwn>
Mon 30 Nov 2009 12:23:57 PM CET

brief problem summary:

I'm having a sort of a "bug" after compiling with winavr 2009-03-13 because i noticed that i wasn't receiving one specific CAN message for some reason. The reason was the msg_id on the transmitter was not correct so the receiver didn't "caught" any because the ID were diferent. But the main problem was that in code C was everything perfect but when I saw the assembly, there was some wrong stuff there (see below please).
The ID is a #define and for some reason r16 is wrongly changed into the last ID byte... making an erroneous ID.

**Comment by avrs-admin** [comment]
Eric Weddington <arcanum>
Wed 13 Jan 2010 06:06:27 AM CET

Please be succinct in describing the problem, and why you think this is a bug in avr-libc. If you think the problem is in the compiler then you need to submit a bug to the GCC project here:
 [http://gcc.gnu.org/](https://savannah.nongnu.orghttp://gcc.gnu.org/)

If you think it is a problem with avr-libc, then please provide a **small** compilable test case that shows the problem, with instructions on how to reproduce the problem.

**Comment by avrs-admin** [comment]
Jan Waclawek <wek>
Wed 13 Jan 2010 09:10:29 AM CET

This bug has been discussed extensively in the avrfreaks.net thread to which the link is given in the original bugreport.

It has been confirmed that while the given code DOES result in erratic binary, reducing the given code results in correct code.

It has also been noted that this **might** be the same error than [http://gcc.gnu.org/bugzilla/show_bug.cgi?id=37466](https://savannah.nongnu.orghttp://gcc.gnu.org/bugzilla/show_bug.cgi?id=37466) , which is said to be fixed in gcc 4.3.3

So, some of the principal developers could perhaps try and confirm that gcc 4.3.3 fixes this bug, too. Then there's no point to submit a duplicate bugreport to an already fixed bug into the gcc bugtracker.

JW

**Comment by avrs-admin** [comment]
Joerg Wunsch <joerg_wunsch>
Wed 09 Jun 2010 11:31:35 AM CEST

The resulting code snippet when compiling the code from the
attachment here using GCC 4.3.4 is:

```
can_tx_msg2.id         = (uint32_t) TEST_CAN_TX_MSG2_ID;
5de:   80 ed           ldi     r24, 0xD0       ; 208
5e0:   e8 2e           mov     r14, r24
5e2:   8f ef           ldi     r24, 0xFF       ; 255
5e4:   f8 2e           mov     r15, r24
5e6:   8f ef           ldi     r24, 0xFF       ; 255
5e8:   08 2f           mov     r16, r24
5ea:   8d ef           ldi     r24, 0xFD       ; 253
5ec:   18 2f           mov     r17, r24
5ee:   e0 92 5f 01     sts     0x015F, r14
5f2:   f0 92 60 01     sts     0x0160, r15
5f6:   00 93 61 01     sts     0x0161, r16
5fa:   10 93 62 01     sts     0x0162, r17

```
So closing this as being fixed in GCC now (it's "invalid"
with respect to the scope of avr-libc anyway, as this has been
a compiler issue).
